llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -1161,6 +1161,14 @@ bool MemCpyOptPass::processMemCpyMemCpyDependence(MemCpyInst *M,
                      MSSA->getMemoryAccess(MDep), MSSA->getMemoryAccess(M)))
     return false;
 
+  // No need to create `memcpy(a <- a)`.
+  if (BAA.isMustAlias(M->getDest(), MDep->getRawSource())) {
```

Comment:
```suggestion
  if (!M->isVolatile() && BAA.isMustAlias(M->getDest(), MDep->getRawSource())) {
```
> The optimizers must not change the number of volatile operations or change their order of execution relative to other volatile operations.

Reply:
It's already checked here: https://github.com/llvm/llvm-project/blob/385ace339a9fc6fcff429d30da5667a25f0a9888/llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp#L1670